src/protocol/modulus_conversion/mod.rs

```diff
@@ -16,17 +16,80 @@ use crate::{ff::Field, secret_sharing::replicated::semi_honest::AdditiveShare as
 ///     ...
 ///     `[ row[n].bit0, row[n].bit1, ..., row[n].bit31 ]`,
 ///  `]`
+/// `num_multi_bits`: L
 ///
 /// output:
 /// `[`
-///     `[ row[0].bit0, row[1].bit0, ..., row[n].bit0 ]`,
-///     `[ row[0].bit1, row[1].bit1, ..., row[n].bit1 ]`,
+///     `[ row[0].bit0, row[0].bit1, ..., row[0].bitL ], [ row[1].bit0, row[1].bit1, ..., row[1].bitL ], .. [ row[n].bit0, row[n].bit1, ..., row[n].bitL ]`,
+///     `[ row[0].bitL+1, ..., row[0].bit2L ], [ row[1].bitL+1, ..., row[1].bit2L ], .. [ row[n].bitL+1, ..., row[n].bit2L ], `,
 ///     ...
-///     `[ row[0].bit31, row[1].bit31, ..., row[n].bit31 ]`,
+///     `[ row[0].bitmL,  ..., row[0].bit31 ], [ row[1].bitmL, ..., row[n].bit31 ], .. [ row[n].bitmL, ..., row[n].bit31 ]`,
 /// `]`
+/// # Errors
+/// If any underlying protocol fails
+/// # Panics
+/// Panics if input doesn't have same number of bits as `num_bits`
 #[must_use]
-pub fn transpose<F: Field>(input: &[Vec<Replicated<F>>]) -> Vec<Vec<Replicated<F>>> {
-    (0..input[0].len())
-        .map(|i| input.iter().map(|b| b[i].clone()).collect::<Vec<_>>())
+pub fn split_into_multi_bit_slices<F: Field>(
+    input: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Vec<Vec<Vec<Replicated<F>>>> {
+    let num_multi_bits = num_multi_bits;
+    let total_records = input.len();
+    (0..num_bits)
+        .step_by(num_multi_bits.try_into().unwrap())
+        .map(|bit_num| {
+            (0..total_records)
+                .map(|record_idx| {
+                    let last_bit_num = std::cmp::min(bit_num + num_multi_bits, num_bits) as usize;
+                    let one_slice = &input[record_idx][bit_num as usize..last_bit_num];
+                    one_slice.to_vec()
+                })
+                .collect::<Vec<_>>()
+        })
         .collect::<Vec<_>>()
 }
+
+/// Combine 3D vectors to get 2D vectors
+///
+/// input:
+/// `[`
+///     `[ row[0].bit0, row[0].bit1, ..., row[0].bitL ], [ row[1].bit0, row[1].bit1, ..., row[1].bitL ], .. [ row[n].bit0, row[n].bit1, ..., row[n].bitL ]`,
+///     `[ row[0].bitL+1, ..., row[0].bit2L ], [ row[1].bitL+1, ..., row[1].bit2L ], .. [ row[n].bitL+1, ..., row[n].bit2L ], `,
+///     ...
+///     `[ row[0].bitmL,  ..., row[0].bit31 ], [ row[1].bitmL, ..., row[n].bit31 ], .. [ row[n].bitmL, ..., row[n].bit31 ]`,
+/// `]`
+/// `num_multi_bits`: L
+///
+/// output:
+/// `[`
+///     `[ row[0].bit0, row[0].bit1, ..., row[0].bit31 ]`,
+///     `[ row[1].bit0, row[1].bit1, ..., row[1].bit31 ]`,
+///     ...
+///     `[ row[n].bit0, row[n].bit1, ..., row[n].bit31 ]`,
+///  `]`
+
+/// # Errors
+/// If any underlying protocol fails
```

Comment:
that does not seem like a correct description

Reply:
This indeed is the right description. we are basically collapsing same record bits together in the same row

Reply:
I meant that you don't return `Result` so there are no errors that can be reported from `combine_slice`

Reply:
Oh my bad.. fixing this